### ISSUE TITLE
support AlternativeImage input in OCR-D processor

### ISSUE THREAD

**Issue by bertsky** [qurator-spk/sbb_textline_detection#32]
The [README](https://github.com/qurator-spk/sbb_textline_detection#usage-with-ocr-d) states the following about current OCR-D support:

> Usage with OCR-D
> 
> ```
> ocrd-example-binarize -I OCR-D-IMG -O OCR-D-IMG-BIN
> ocrd-sbb-textline-detector -I OCR-D-IMG-BIN -O OCR-D-SEG-LINE-SBB \
>         -p '{ "model": "/path/to/the/models/textline_detection" }'
> ```
> 
> Segmentation works on raw RGB images, but respects and retains AlternativeImages from binarization steps, so it's a good idea to do binarization first, then perform the textline detection. The used binarization processor must produce an AlternativeImage for the binarized image, not replace the original raw RGB image.

That latter statement shows that there has been some awareness of the problem, but unfortunately it is not consistent with the usage example given above, or with the actual implementation.

IIUC All the implementation (and the usage example) does is take a binarized image _as if it were the original image_ and create a segmentation for it. It does not even look at the `AlternativeImage`s of the PAGE input file. This will **not work** for OCR-D pipelines where that binarized image has also been cropped (due to a prior `Border` detection) or deskewed (due to a prior `@orientation` detection) or dewarped. The image itself cannot tell you, only the PAGE-XML's `AlternativeImage/@comments` can.

So please use `workspace.image_from_page` instead of bare `input_file.local_filename` or `page.imageFilename`. (The latter two also would not work for remote images.)

**Comment by mikegerber** [comment]
It 

1. uses the original image 
2. retains the `AlternativeImage` for use in a subsequent OCR step.

I do not know how this behaves if anything was cropped before though.

> IIUC All the implementation (and the usage example) does is take a binarized image as if it were the original image and create a segmentation for it. 

No. What it does is taking the *original RGB* image from the PAGE file that was generated by the binarization step, operates on that and retains the (unused) binarized AlternativeImages in the output.

> So please use workspace.image_from_page instead of bare input_file.local_filename or page.imageFilename. (The latter two also would not work for remote images.)

Does this give me the original RGB image?

**Comment by bertsky** [comment]
> It
> 
>     1. uses the original image
>     2. retains the `AlternativeImage` for use in a subsequent OCR step.

Which is very different from what is stated above: `but respects and retains AlternativeImages from binarization steps, so it's a good idea to do binarization first`. _Respecting_ AlternativeImage and even demanding their use (for binarization) is exactly the opposite of just passing them through and ignoring them otherwise.

> > IIUC All the implementation (and the usage example) does is take a binarized image as if it were the original image and create a segmentation for it.
> 
> No. What it does is taking the _original RGB_ image from the PAGE file that was generated by the binarization step, operates on that and retains the (unused) binarized AlternativeImages in the output.

That's exactly what `as if it were the original image` is. There can only be **one original** you know, so let's be clear about the terms we use.

> > So please use workspace.image_from_page instead of bare input_file.local_filename or page.imageFilename. (The latter two also would not work for remote images.)
> 
> Does this give me the original RGB image?

If you want to. See `feature_selector` kwarg. Usage examples in virtually every existing processor that deals with images.

**Comment by mikegerber** [comment]
> > 1. uses the original image
> > 2. retains the `AlternativeImage` for use in a subsequent OCR step.
> Which is very different from what is stated above: `but respects and retains AlternativeImages from binarization steps, so it's a good idea to do binarization first`. _Respecting_ AlternativeImage and even demanding their use (for binarization) is exactly the opposite of just passing them through and ignoring them otherwise.

Yeah "respecting" is probably the incorrect term, I changed that.

I'd also ask you to please review your tone in your comments. Your comments often come across as 
bossy.

> > > So please use workspace.image_from_page instead of bare input_file.local_filename or page.imageFilename. (The latter two also would not work for remote images.)
> > Does this give me the original RGB image?
> If you want to. See `feature_selector` kwarg. Usage examples in virtually every existing processor that deals with images.

We need the RGB image aka "the original image", that's the whole point of not using the AlternativeImages and that whole arguably ambiguous (or even incorrect) paragraph of the README.

I am going to change that part to get the RGB image using `image_from_page`.

**Comment by bertsky** [comment]
> I'd also ask you to please review your tone in your comments. Your comments often come across as
> bossy.

Huh? Where, how?

> We need the RGB image aka "the original image", that's the whole point of not using the AlternativeImages and that whole arguably ambiguous (or even incorrect) paragraph of the README.

Oh, I see. That's the source of our confusion. You equated _original_ with _raw_, for me (and for OCR-D) it's the _unmodified_ first image that enters the pipeline (and is still the reference when you leave it), regardless of whether it's RGB, grayscale, already binarized etc.

Your use-case is covered exactly by the AlternativeImage handling in the OCR-D spec and the implementation in core. That's what `@comments` (or _image features_) are used for, and that's what `feature_filter` / `feature_selector` kwargs are doing.

> I am going to change that part to get the RGB image using `image_from_page`.

Splendid. Have a look at ocrd-cis-ocropy-binarize for an example of using `feature_filter='binarized'` or at ocrd-tesserocr-segment-region for an example of writing back correct absolute coordinates via `coordinates_for_segment`.

**Comment by mikegerber** [comment]
Side note to self: I can absolutely not remember why I recommended running binarization before this segmentation. I'll investigate.

Another thing that lead to the somewhat confusing README text: The original version just wrote a new PAGE file, then I reworked it to merge results into a existing PAGE to support ocrd_olena output.

**Comment by mikegerber** [comment]
Should be fixed in 020ffbc.

**Comment by bertsky** [comment]
> Should be fixed in [020ffbc](https://github.com/qurator-spk/sbb_textline_detection/commit/020ffbcfde0c09ba9f0221f5a136f821fb0c1281).

Thanks – that's much better, but not quite correct yet: You need to re-base the detected `Border` on the input image's coordinate system, too! (And this coord system can not only differ due to cropping earlier in the workflow, but also deskewing.)

Also, for this tool it's probably better to process on the _uncropped_ image. So instead of just `self.workspace.image_from_page(page, page_id, feature_filter='binarized,grayscale_normalized')`, use `feature_filter='cropped,binarized,grayscale_normalized'`. Unless your detector is not good at cropping on its own (which I doubt), this will also yield better quality (because cropping is easier on uncropped images than on badly cropped images).

Note that then you would _still_ have to rebase all coords, because you did not rule out `'deskewed'` yet. (Think `ocrorot`.)

To summarise: re-transform Border coords as well, and use uncropped input images.

Besides, the incriminated formulation in the README is still up there :-p

**Comment by mikegerber** [comment]
- [x] use uncropped input images: 006c776
- [x] re-transform Border coords: 665b739
  - CoordsType vs points vs polygons could be a bit more convenient.
- "the incriminated formulation": Please open a PR, I'm sure @vahidrezanezhad or @cneud will merge it (I'm not maintaining the OCR-D interface here anymore, this is the last issue I'm handling)

**Comment by bertsky** [comment]
> * CoordsType vs points vs polygons could be a bit more convenient.

Yes, we could have some convenience functions for Coords/points – in ocrd_utils or directly in the PAGE class as getter/setter. Feel free to open an issue in core if you have a concrete idea.

> * [x]  use uncropped input images: [006c776](https://github.com/qurator-spk/sbb_textline_detection/commit/006c7765b08aa61e1ea8f0912dcdc3b3e02bf499)
> * [x]  re-transform Border coords: [665b739](https://github.com/qurator-spk/sbb_textline_detection/commit/665b739fb840c6ca8cf97ba3999726662a42867b)

Excellent!

> * "the incriminated formulation": Please open a PR, I'm sure @vahidrezanezhad or @cneud will merge it (I'm not maintaining the OCR-D interface here anymore, this is the last issue I'm handling)

Will do. Thanks again.
